### PR TITLE
[WIP] Add dev container support

### PR DESCRIPTION
**Pull request by Huoleit**
In this PR, dev container support is added. The whole dev container is split into two parts. 
- `.github/.devcontainer`: holds everything needed for the SMB simulation environment. It should be built and uploaded to the GitHub registry by CI. For users who wish to run commands with docker CLI, it is also the image pulled from the register.
- `.devcontainer`: It is a top-level dev config that is dedicated to vscode devcontainer extension. It extends the base image and allows users to customize their development environment. For example, add vscode extensions, preserve bash history, etc.

More details [here](https://onyx-time-1bc.notion.site/Apr-10-2f4f4c13aa624b14a4616886e15601db?pvs=4)

## TODO
- [x] Add base .devcontainer.json and Dockerfile
- [x] Add cache-preserving multi-stage build to the base image
- [x] Add VNC support - Windows and Mac users can fallback to VNC if X11 forwarding does not work 
- [ ] Add `postCreateCommand.bash` to compile catkin_ws
- [ ] Add development tools and vscode extensions
- [ ] Add alias
- [ ] Finalize top-level .devcontainer.json and Dockerfile
- [ ] Add dev container CI
- [ ] Shrink image size (the whole image is 13GB, Pytorch takes 8GB)

### CODE REVIEW

**Review by mantelt — COMMENTED** [review]
First of all: thanks for looking into this!
I'm new to dev containers, so please be patient ;-) 

As pointed out before, I would like to achieve two goals:

- have some basic CI to make sure everything builds. currently, this is achieved by building smb_dev and SuperMegaBot using Github Actions. 

- distribute a dev environment to the participants that is easy to set up. (I referred to this as deployment, but only now figured out that this might be bad usage of the term)

From what I gathered so far, a dev container is the right way to achieve both goals in one go. But we might need to adjust a few github actions to make everything consistent and keep it efficient.

To me, it feels like it will be more efficient if we quickly discuss your idea on what to implement once more directly. What do you think?

**Comment by Huoleit** [comment]
@deepanaishtaweera  and I had a meeting about the whole setup a week before. More details are shown in the notes here.
https://onyx-time-1bc.notion.site/Apr-10-2f4f4c13aa624b14a4616886e15601db?pvs=4

File: .devcontainer/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,9 @@
+# syntax=docker/dockerfile:1
+
+FROM ghcr.io/ETHZ-RobotX/smb_docker
```
[the remark this answers is not in the document]
Comment (Huoleit) [reply .devcontainer/Dockerfile:3]: Top-level `.devcontainer` i.e. this file is not finalized yet as indicated in the TODO section. The base image will be built once I finish the ci part. And I will replace this line to the correct image name.

File: .github/.devcontainer/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+ARG OVERLAY_WS
+ARG SMB_RAW_REPO_URL=https://raw.githubusercontent.com/ETHZ-RobotX/SuperMegaBot/master
+
+RUN DEBIAN_FRONTEND=noninteractive apt-get update && apt-get install -y --no-install-recommends \
+    python3-vcstool \
+    git
+
+WORKDIR ${OVERLAY_WS}/src
+RUN vcs import --recursive --input ${SMB_RAW_REPO_URL}/smb.repos && \
+    vcs import --recursive --input ${SMB_RAW_REPO_URL}/smb_hw.repos
```
[the remark this answers is not in the document]
Comment (Huoleit) [reply .github/.devcontainer/Dockerfile:17]: Both `smb.repos` and `smb_hw.repos` are imported in the old Dockerfile. I simply copy them here without checking the dependencies. If `smb.repos` does not depend on `smb_hw.repos` and `smb.repos` alone is enough to run the simulation, we can safely remove `smb_hw.repos`.

**Comment by Huoleit** [comment]
> First of all: thanks for looking into this! I'm new to dev containers, so please be patient ;-)
> 
> As pointed out before, I would like to achieve two goals:
> 
> * have some basic CI to make sure everything builds. currently, this is achieved by building smb_dev and SuperMegaBot using Github Actions.
> * distribute a dev environment to the participants that is easy to set up. (I referred to this as deployment, but only now figured out that this might be bad usage of the term)
> 
> From what I gathered so far, a dev container is the right way to achieve both goals in one go. But we might need to adjust a few github actions to make everything consistent and keep it efficient.
> 
> To me, it feels like it will be more efficient if we quickly discuss your idea on what to implement once more directly. What do you think?

Sure, we can discuss it and I can give a demo. I'll message you on Slack to schedule a time.

**Pull request closed without merging**